framework/resource/resource.go

```diff
@@ -0,0 +1,443 @@
+// WithState. If WithState was never called, this is guaranteed to always
+// be nil.
+//
+// The returned value is also a direct pointer to the internally stored
+// state so it should not be modified simultaneously to any resource
+// operations.
+func (r *Resource) State() interface{} {
+	return r.stateValue
+}
+
+// SetState manually sets the state for this resource. This is not recommended;
+// typically state should only be modified through the lifecycle functions.
+// However, this function is particularly useful to transition to using the
+// resource manager from a previous version that didn't use the resource manager.
+//
+// The value v must be the same type as the type given for WithState.
+func (r *Resource) SetState(v interface{}) error {
+	if reflect.TypeOf(v) != r.stateType {
+		return fmt.Errorf("state value type %T does not match expected type %s",
+			v, r.stateType.String())
+	}
+
+	r.stateValue = v
+	return nil
+}
+
+// Create creates this resource. args is a list of arguments to make
+// available to the creation function via dependency injection (matching
+// types in the arguments).
+//
+// After Create is called, any state can be accessed via the State function.
+// This may be populated even during failure with partial state.
+func (r *Resource) Create(args ...interface{}) error {
+	f, err := r.mapperForCreate(nil)
+	if err != nil {
+		return err
+	}
+
+	mapperArgs := make([]argmapper.Arg, len(args))
+	for i, v := range args {
+		mapperArgs[i] = argmapper.Typed(v)
+	}
+
+	result := f.Call(mapperArgs...)
+	return result.Err()
+}
+
+// Destroy destroys this resource. args is a list of arguments to make
+// available to the destroy function via dependency injection. The state
+// value will always be available as an argument (though it may be nil
+// if this resource has no state).
+//
+// After Destroy is called successfully (without an error result), the
+// state will always be nil.
+func (r *Resource) Destroy(args ...interface{}) error {
+	f, err := r.mapperForDestroy(nil)
+	if err != nil {
+		return err
+	}
+
+	mapperArgs := make([]argmapper.Arg, len(args))
+	for i, v := range args {
+		mapperArgs[i] = argmapper.Typed(v)
+	}
+
+	result := f.Call(mapperArgs...)
+	return result.Err()
+}
+
+// mapperForCreate returns an argmapper func that takes as input the
+// requirements for the createFunc and returns the state type plus an error.
+// This creates a valid "mapper" we can use with Manager.
+func (r *Resource) mapperForCreate(cs *createState) (*argmapper.Func, error) {
+	// Create the func for the createFunc as-is. We need to get the input/output sets.
+	original, err := argmapper.NewFunc(r.createFunc)
+	if err != nil {
+		return nil, err
+	}
+
+	// For our output, we will always output our unique marker type.
+	// This unique marker type will allow our resource manager to create
+	// a function chain that calls all the resources necessary.
+	markerVal := markerValue(r.name)
```

Comment:
Either we need to autogenerate `.name` if it's empty or it shouldn't be an option and instead a required argument passed to `NewResource`.

Reply:
I actually forgot to implement the `Validate` function which indeed should require the name. The validate function will also check all the arg mappings and so on so that it is viable...

Reply:
If it's required, should it just be an argument to NewResource?

Reply:
A create func is also required, and there are other stipulations that must be verified, so I wanted to use validation funcs instead (similar to TF SDK).

Reply:
Added validation!